src/logger.coffee

```diff
@@ -5,114 +5,237 @@ es = require 'event-stream'
 Lock = require 'rwlock'
 { checkTruthy } = require './lib/validation'
 
-LOG_PUBLISH_INTERVAL = 110
+class NativeLoggerBackend
+	constructor: ->
+		@logPublishInterval = 1000
+		@maxLogLinesPerBatch = 10
+		@maxQueueSize = 100
+		@maxLineLength = 300
+		@publishQueue = []
+		@intervalHandle = null
+		@publishEnabled = false
+		@offlineMode = false
+		@shouldStop = false
+
+	publishLoop: =>
+		startTime = process.hrtime()
+		Promise.try =>
+			return if @offlineMode or !@publishEnabled or @publishQueue.length is 0
+			currentBatch = _.slice(@publishQueue, 0, @maxLogLinesPerBatch)
+			@publishQueue = _.slice(@publishQueue, @maxLogLinesPerBatch)
+			# Silently ignore errors sending logs (just like with pubnub)
+			@apiBinder.logBatch(currentBatch)
+		.catchReturn()
+		.then =>
+			elapsedTime = process.hrtime(startTime)
+			elapsedTimeMs = elapsedTime[0] * 1000 + elapsedTime[1] / 1e6
+			nextDelay = Math.min(@logPublishInterval - elapsedTimeMs, 0)
+			Promise.delay(nextDelay)
+		.then =>
+			if @shouldStop
+				return
+			@publishLoop()
 
-# Pubnub's message size limit is 32KB (unclear on whether it's KB or actually KiB,
-# but we'll be conservative). So we limit a log message to 2 bytes less to account
-# for the [ and ] in the array.
-MAX_LOG_BYTE_SIZE = 30000
-MAX_MESSAGE_INDEX = 9
+	_truncateToMaxLength: (message) =>
+		if message.message.length > @maxLineLength
+			message = _.clone(message)
+			message.message = _.truncate(message.message, { length: @maxLineLength, omission: '[...]' })
+		return message
 
-module.exports = class Logger
-	constructor: ({ @eventTracker }) ->
+	logDependent: (msg, { uuid }) =>
+		msg = @_truncateToMaxLength(msg)
+		@apiBinder.logDependent(uuid, msg)
+
+	log: (message) =>
+		return if @offlineMode or !@publishEnabled or @publishQueue.length >= @maxQueueSize
+		if _.isString(message)
+			message = { message: message }
+
+		message = _.defaults({}, message, {
+			timestamp: Date.now()
+			message: ''
+		})
+
+		message = @_truncateToMaxLength(message)
+
+		if @publishQueue.length < @maxQueueSize - 1
+			@publishQueue.push(_.pick(message, [ 'message', 'timestamp', 'isSystem', 'isStderr', 'imageId' ]))
+		else
+			@publishQueue.push({
+				message: 'Warning! Some logs dropped due to high load'
+				timestamp: Date.now()
+				isSystem: true
```

Comment:
We changed the log properties to underscore (in one of the arch calls). It's what the api expects, check the [API tests](https://github.com/resin-io/resin-api/blob/master/test/24_device_log.coffee#L34) when in doubt

Reply:
Yes, but internally we use camel case everywhere. The APIBinder converts to snake in `_sendLogsRequest`